File: app/src/main/java/awais/instagrabber/fragments/directmessages/DirectMessageInboxFragment.java
package awais.instagrabber.fragments.directmessages;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.IntentFilter;
import android.content.res.Configuration;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.coordinatorlayout.widget.CoordinatorLayout;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.lifecycle.ViewModelStoreOwner;
import androidx.navigation.NavController;
import androidx.navigation.NavDirections;
import androidx.navigation.fragment.NavHostFragment;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.badge.BadgeDrawable;
import com.google.android.material.badge.BadgeUtils;
import com.google.android.material.bottomnavigation.BottomNavigationView;

import java.util.List;

import awais.instagrabber.R;
import awais.instagrabber.activities.MainActivity;
import awais.instagrabber.adapters.DirectMessageInboxAdapter;
import awais.instagrabber.broadcasts.DMRefreshBroadcastReceiver;
import awais.instagrabber.customviews.helpers.RecyclerLazyLoaderAtEdge;
import awais.instagrabber.databinding.FragmentDirectMessagesInboxBinding;
import awais.instagrabber.repositories.responses.directmessages.DirectThread;
import awais.instagrabber.viewmodels.DirectInboxViewModel;

public class DirectMessageInboxFragment extends Fragment implements SwipeRefreshLayout.OnRefreshListener {
    private static final String TAG = "DirectMessagesInboxFrag";

    private CoordinatorLayout root;
    private RecyclerLazyLoaderAtEdge lazyLoader;
    private DirectInboxViewModel viewModel;
    private boolean shouldRefresh = true;
    private FragmentDirectMessagesInboxBinding binding;
    private DMRefreshBroadcastReceiver receiver;
    private DirectMessageInboxAdapter inboxAdapter;
    private MainActivity fragmentActivity;
    private boolean scrollToTop = false;
    private boolean navigating;
    private Observer<List<DirectThread>> threadsObserver;
    private MenuItem pendingRequestsMenuItem;
    private BadgeDrawable pendingRequestTotalBadgeDrawable;
    private boolean isPendingRequestTotalBadgeAttached;

    @Override
    public void onCreate(@Nullable final Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fragmentActivity = (MainActivity) getActivity();
        if (fragmentActivity != null) {
            final NavController navController = NavHostFragment.findNavController(this);
            final ViewModelStoreOwner viewModelStoreOwner = navController.getViewModelStoreOwner(R.id.direct_messages_nav_graph);
            viewModel = new ViewModelProvider(viewModelStoreOwner).get(DirectInboxViewModel.class);
        }
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(@NonNull final LayoutInflater inflater,
                             final ViewGroup container,
                             final Bundle savedInstanceState) {
        if (root != null) {
            shouldRefresh = false;
            return root;
        }
        binding = FragmentDirectMessagesInboxBinding.inflate(inflater, container, false);
        root = binding.getRoot();
        return root;
    }

    @Override
    public void onViewCreated(@NonNull final View view, @Nullable final Bundle savedInstanceState) {
        if (!shouldRefresh) return;
        init();
    }

    @Override
    public void onRefresh() {
        lazyLoader.resetState();
        scrollToTop = true;
        if (viewModel != null) {
            viewModel.refresh();
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        unregisterReceiver();
    }

    @Override
    public void onResume() {
        super.onResume();
        setupObservers();
        final Context context = getContext();
        if (context == null) return;
        receiver = new DMRefreshBroadcastReceiver(() -> {
            Log.d(TAG, "onResume: broadcast received");
            // refreshInbox = true;
        });
        context.registerReceiver(receiver, new IntentFilter(DMRefreshBroadcastReceiver.ACTION_REFRESH_DM));
    }

    @SuppressLint("UnsafeExperimentalUsageError")
    @Override
    public void onDestroyView() {
        super.onDestroyView();
        unregisterReceiver();
        isPendingRequestTotalBadgeAttached = false;
        if (pendingRequestTotalBadgeDrawable != null) {
            BadgeUtils.detachBadgeDrawable(pendingRequestTotalBadgeDrawable, fragmentActivity.getToolbar(), pendingRequestsMenuItem.getItemId());
            pendingRequestTotalBadgeDrawable = null;
        }
    }

    @Override
    public void onCreateOptionsMenu(@NonNull final Menu menu, @NonNull final MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        pendingRequestsMenuItem = menu.add(Menu.NONE, R.id.pending_requests, Menu.NONE, "Pending requests");
        pendingRequestsMenuItem.setIcon(R.drawable.ic_account_clock_24)
                               .setVisible(false)
                               .setShowAsActionFlags(MenuItem.SHOW_AS_ACTION_ALWAYS);
        attachPendingRequestsBadge(viewModel.getPendingRequestsTotal().getValue());
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull final MenuItem item) {
        if (item.getItemId() == R.id.pending_requests) {
            final NavDirections directions = DirectMessageInboxFragmentDirections.actionInboxToPendingInbox();
            NavHostFragment.findNavController(this).navigate(directions);
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    private void unregisterReceiver() {
        if (receiver == null) return;
        final Context context = getContext();
        if (context == null) return;
        context.unregisterReceiver(receiver);
        receiver = null;
    }

    @Override
    public void onConfigurationChanged(@NonNull final Configuration newConfig) {
        super.onConfigurationChanged(newConfig);
        init();
    }

    @Override
    public void onDestroy() {
        super.onDestroy();
        removeViewModelObservers();
        viewModel.onDestroy();
    }

    private void setupObservers() {
        threadsObserver = list -> {
            if (inboxAdapter == null) return;
            inboxAdapter.submitList(list, () -> {
                if (!scrollToTop) return;
                binding.inboxList.smoothScrollToPosition(0);
                scrollToTop = false;
            });
        };
        viewModel.getThreads().observe(fragmentActivity, threadsObserver);
        viewModel.getFetchingInbox().observe(getViewLifecycleOwner(), fetching -> binding.swipeRefreshLayout.setRefreshing(fetching));
        viewModel.getUnseenCount().observe(getViewLifecycleOwner(), this::setBottomNavBarBadge);
        viewModel.getPendingRequestsTotal().observe(getViewLifecycleOwner(), this::attachPendingRequestsBadge);
    }

    @SuppressLint("UnsafeExperimentalUsageError")
    private void attachPendingRequestsBadge(@Nullable final Integer count) {
        if (pendingRequestsMenuItem == null) return;
        if (pendingRequestTotalBadgeDrawable == null) {
            final Context context = getContext();
            if (context == null) return;
            pendingRequestTotalBadgeDrawable = BadgeDrawable.create(context);
        }
        if (count == null || count == 0) {
            BadgeUtils.detachBadgeDrawable(pendingRequestTotalBadgeDrawable, fragmentActivity.getToolbar(), pendingRequestsMenuItem.getItemId());
            isPendingRequestTotalBadgeAttached = false;
            pendingRequestTotalBadgeDrawable.setNumber(0);
            pendingRequestsMenuItem.setVisible(false);
            return;
        }
        pendingRequestsMenuItem.setVisible(true);
        if (pendingRequestTotalBadgeDrawable.getNumber() == count) return;
        pendingRequestTotalBadgeDrawable.setNumber(count);
        if (!isPendingRequestTotalBadgeAttached) {
            BadgeUtils.attachBadgeDrawable(pendingRequestTotalBadgeDrawable, fragmentActivity.getToolbar(), pendingRequestsMenuItem.getItemId());
            isPendingRequestTotalBadgeAttached = true;
        }
    }

    private void removeViewModelObservers() {
        if (viewModel == null) return;
        if (threadsObserver != null) {
            viewModel.getThreads().removeObserver(threadsObserver);
        }
        // no need to explicitly remove observers whose lifecycle owner is getViewLifecycleOwner
    }

    private void init() {
        final Context context = getContext();
        if (context == null) return;
        setupObservers();
        binding.swipeRefreshLayout.setOnRefreshListener(this);
        binding.inboxList.setHasFixedSize(true);
        binding.inboxList.setItemViewCacheSize(20);
        final LinearLayoutManager layoutManager = new LinearLayoutManager(context);
        binding.inboxList.setLayoutManager(layoutManager);
        inboxAdapter = new DirectMessageInboxAdapter(thread -> {
            if (navigating) return;
            navigating = true;
            final Bundle bundle = new Bundle();
            bundle.putString("threadId", thread.getThreadId());
            bundle.putString("title", thread.getThreadTitle());
            if (isAdded()) {
                NavHostFragment.findNavController(this).navigate(R.id.action_inbox_to_thread, bundle);
            }
            navigating = false;
        });
        inboxAdapter.setHasStableIds(true);
        binding.inboxList.setAdapter(inboxAdapter);
        lazyLoader = new RecyclerLazyLoaderAtEdge(layoutManager, page -> {
            if (viewModel == null) return;
            viewModel.fetchInbox();
        });
        binding.inboxList.addOnScrollListener(lazyLoader);
    }

    private void setBottomNavBarBadge(final int unseenCount) {
        final BottomNavigationView bottomNavView = fragmentActivity.getBottomNavView();
        final BadgeDrawable badge = bottomNavView.getOrCreateBadge(R.id.direct_messages_nav_graph);
        if (badge == null) return;
        if (unseenCount == 0) {
            badge.setVisible(false);
            badge.clearNumber();
            return;
        }
        if (badge.getVerticalOffset() != 10) {
            badge.setVerticalOffset(10);
        }
        badge.setNumber(unseenCount);
        badge.setVisible(true);
    }
}

File: app/src/main/java/awais/instagrabber/fragments/directmessages/DirectMessageInboxFragment.kt
package awais.instagrabber.fragments.directmessages

import android.annotation.SuppressLint
import android.content.res.Configuration
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.util.Log
import android.view.*
import androidx.fragment.app.Fragment
import androidx.fragment.app.activityViewModels
import androidx.navigation.fragment.findNavController
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout.OnRefreshListener
import awais.instagrabber.R
import awais.instagrabber.activities.MainActivity
import awais.instagrabber.adapters.DirectMessageInboxAdapter
import awais.instagrabber.customviews.helpers.RecyclerLazyLoaderAtEdge
import awais.instagrabber.databinding.FragmentDirectMessagesInboxBinding
import awais.instagrabber.models.Resource
import awais.instagrabber.repositories.responses.directmessages.DirectInbox
import awais.instagrabber.repositories.responses.directmessages.DirectThread
import awais.instagrabber.utils.extensions.TAG
import awais.instagrabber.viewmodels.DirectInboxViewModel
import com.google.android.material.badge.BadgeDrawable
import com.google.android.material.badge.BadgeUtils
import com.google.android.material.internal.ToolbarUtils
import com.google.android.material.snackbar.Snackbar

class DirectMessageInboxFragment : Fragment(), OnRefreshListener {
    private val viewModel: DirectInboxViewModel by activityViewModels()

    private lateinit var fragmentActivity: MainActivity
    private lateinit var binding: FragmentDirectMessagesInboxBinding
    private lateinit var lazyLoader: RecyclerLazyLoaderAtEdge

    private var scrollToTop = false
    private var navigating = false

    private var pendingRequestsMenuItem: MenuItem? = null
    private var pendingRequestTotalBadgeDrawable: BadgeDrawable? = null
    private var isPendingRequestTotalBadgeAttached = false
    private var inboxAdapter: DirectMessageInboxAdapter? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        fragmentActivity = requireActivity() as MainActivity
        setHasOptionsMenu(true)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?,
    ): View {
        binding = FragmentDirectMessagesInboxBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        init()
    }

    override fun onRefresh() {
        lazyLoader.resetState()
        scrollToTop = true
        viewModel.refresh()
    }

    @SuppressLint("UnsafeExperimentalUsageError", "UnsafeOptInUsageError", "RestrictedApi")
    override fun onPause() {
        super.onPause()
        isPendingRequestTotalBadgeAttached = false
        pendingRequestsMenuItem?.let {
            val menuItemView = ToolbarUtils.getActionMenuItemView(fragmentActivity.toolbar, it.itemId)
            if (menuItemView != null) {
                BadgeUtils.detachBadgeDrawable(pendingRequestTotalBadgeDrawable, fragmentActivity.toolbar, it.itemId)
                pendingRequestTotalBadgeDrawable = null
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        inflater.inflate(R.menu.dm_inbox_menu, menu)
        pendingRequestsMenuItem = menu.findItem(R.id.pending_requests)
        pendingRequestsMenuItem?.isVisible = isPendingRequestTotalBadgeAttached
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == R.id.pending_requests) {
            try {
                val directions = DirectMessageInboxFragmentDirections.actionToPendingInbox()
                findNavController().navigate(directions)
            } catch (e: Exception) {
                Log.e(TAG, "onOptionsItemSelected: ", e)
            }
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onConfigurationChanged(newConfig: Configuration) {
        super.onConfigurationChanged(newConfig)
        init()
    }

    private fun setupObservers() {
        viewModel.threads.observe(viewLifecycleOwner, { list: List<DirectThread?> ->
            inboxAdapter?.submitList(list) {
                if (!scrollToTop) return@submitList
                binding.inboxList.post { binding.inboxList.smoothScrollToPosition(0) }
                scrollToTop = false
            }
        })
        viewModel.inbox.observe(viewLifecycleOwner, { inboxResource: Resource<DirectInbox?>? ->
            if (inboxResource == null) return@observe
            when (inboxResource.status) {
                Resource.Status.SUCCESS -> binding.swipeRefreshLayout.isRefreshing = false
                Resource.Status.ERROR -> {
                    if (inboxResource.message != null) {
                        Snackbar.make(binding.root, inboxResource.message, Snackbar.LENGTH_LONG).show()
                    }
                    if (inboxResource.resId != 0) {
                        Snackbar.make(binding.root, inboxResource.resId, Snackbar.LENGTH_LONG).show()
                    }
                    binding.swipeRefreshLayout.isRefreshing = false
                }
                Resource.Status.LOADING -> binding.swipeRefreshLayout.isRefreshing = true
            }
        })
        viewModel.pendingRequestsTotal.observe(viewLifecycleOwner, { count: Int? -> attachPendingRequestsBadge(count) })
    }

    @SuppressLint("UnsafeExperimentalUsageError", "UnsafeOptInUsageError", "RestrictedApi")
    private fun attachPendingRequestsBadge(count: Int?) {
        val pendingRequestsMenuItem1 = pendingRequestsMenuItem
        if (pendingRequestsMenuItem1 == null) {
            val handler = Handler(Looper.getMainLooper())
            handler.postDelayed({ attachPendingRequestsBadge(count) }, 500)
            return
        }
        if (pendingRequestTotalBadgeDrawable == null) {
            val context = context ?: return
            pendingRequestTotalBadgeDrawable = BadgeDrawable.create(context)
        }
        if (count == null || count == 0) {
            val menuItemView = ToolbarUtils.getActionMenuItemView(
                fragmentActivity.toolbar,
                pendingRequestsMenuItem1.itemId
            )
            if (menuItemView != null) {
                BadgeUtils.detachBadgeDrawable(pendingRequestTotalBadgeDrawable, fragmentActivity.toolbar, pendingRequestsMenuItem1.itemId)
            }
            isPendingRequestTotalBadgeAttached = false
            pendingRequestTotalBadgeDrawable?.number = 0
            pendingRequestsMenuItem1.isVisible = false
            return
        }
        pendingRequestsMenuItem1.isVisible = true
        if (pendingRequestTotalBadgeDrawable?.number == count) return
        pendingRequestTotalBadgeDrawable?.number = count
        if (!isPendingRequestTotalBadgeAttached) {
            pendingRequestTotalBadgeDrawable?.let {
                BadgeUtils.attachBadgeDrawable(it, fragmentActivity.toolbar, pendingRequestsMenuItem1.itemId)
                isPendingRequestTotalBadgeAttached = true
            }
        }
    }

    private fun init() {
        val context = context ?: return
        setupObservers()
        binding.swipeRefreshLayout.setOnRefreshListener(this)
        binding.inboxList.setHasFixedSize(true)
        binding.inboxList.setItemViewCacheSize(20)
        val layoutManager = LinearLayoutManager(context)
        binding.inboxList.layoutManager = layoutManager
        inboxAdapter = DirectMessageInboxAdapter { thread ->
            val threadId = thread.threadId
            val threadTitle = thread.threadTitle
            if (navigating || threadId.isNullOrBlank() || threadTitle.isNullOrBlank()) return@DirectMessageInboxAdapter
            navigating = true
            if (isAdded) {
                try {
                    val directions = DirectMessageInboxFragmentDirections.actionToThread(threadId, threadTitle)
                    findNavController().navigate(directions)
                } catch (e: Exception) {
                    Log.e(TAG, "init: ", e)
                }
            }
            navigating = false
        }.also {
            it.setHasStableIds(true)
        }
        binding.inboxList.adapter = inboxAdapter
        lazyLoader = RecyclerLazyLoaderAtEdge(layoutManager) { viewModel.fetchInbox() }.also {
            binding.inboxList.addOnScrollListener(it)
        }
    }
}